leptonica: add version 1.83.0

theirix opened the pull request:
Specify library name and version:  **leptonica/1.83.0**

There is one moment I don't fully understand. Upstream uses `find_package(GIF)` which seems to be totally fine.
But Conan generators create a bunch of `giflib-*` files so the find_package doesn't find them but picks up a system include path `/Library/Frameworks/UnixImageIO.framework/Headers`. Using `find_package(GIFLIB)` fixes the problem.

Recent giflib changes do not change a `cmake_file_name` property. If it is not changed, why doesn't it work?


---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

SpaceIm commented:
> There is one moment I don't fully understand. Upstream uses `find_package(GIF)` which seems to be totally fine. But Conan generators create a bunch of `giflib-*` files so the find_package doesn't find them but picks up a system include path `/Library/Frameworks/UnixImageIO.framework/Headers`. Using `find_package(GIFLIB)` fixes the problem.
> 
> Recent giflib changes do not change a `cmake_file_name` property. If it is not changed, why doesn't it work?

Can't reproduce this issue with conan 1.55.0

SpaceIm commented on recipes/leptonica/all/conandata.yml:
```diff
@@ -17,3 +20,5 @@ sources:
   "1.78.0":
     - patch_file: "patches/fix-find-modules-variables.patch"
+      patch_description: "CMake: robust handling of dependencies"
+      patch_type: "conan"
```
```suggestion
      patch_type: "portability"
      patch_source: "https://github.com/DanBloomberg/leptonica/pull/456"
```

theirix commented:
> There is one moment I don't fully understand. Upstream uses find_package(GIF) which seems to be totally fine. But Conan generators create a bunch of giflib-* files so the find_package doesn't find them but picks up a system include path /Library/Frameworks/UnixImageIO.framework/Headers. Using find_package(GIFLIB) fixes the problem.

Removed the local cache, `FindGIF.cmake` appeared and now GIF can be found by `find_package` in the default module mode. False alarm, removed workaround.

SpaceIm commented on recipes/leptonica/all/conanfile.py:
```diff
@@ -85,7 +76,11 @@ def requirements(self):
             self.requires("libtiff/4.4.0")
         if self.options.with_openjpeg:
-            self.requires("openjpeg/2.5.0")
+            if Version(self.version) >= "1.83.0":
+                # newer leptonica can autodetect any openjpeg version
+                self.requires("openjpeg/2.5.0")
+            else:
+                self.requires("openjpeg/2.4.0")
```
It's quite surprising. It was working before this PR with openjpeg 2.5.0 regardless of leptonica version

theirix replied:
I wondered too. Seems leptonica silently ignored it.

Still fighting with detecting it.

SpaceIm replied:
I think it can be reverted to openjpeg 2.5.0 unconditionally now? Still don't understand this new branching.

theirix replied:
Agree, brought it back because it depended on pkgconfig resolution in pre-1.83.0 Leptonica.

Strange build error btw.

SpaceIm commented on recipes/leptonica/all/conanfile.py:
```diff
@@ -139,47 +137,66 @@ def _patch_sources(self):
-            replace_in_file(self, cmake_configure, "if (JP2K_FOUND)", "if(0)")
+        replace_in_file(self, cmakelists_src, "${JP2K_LIBRARIES}", "openjp2")
+        if Version(self.version) < "1.83.0":
+            # pkgconfig is prefered to CMake. Disable pkgconfig so only CMake is used
+            replace_in_file(self, cmakelists, "pkg_check_modules(JP2K libopenjp2>=2.0 QUIET)", "")
```
Surprising. This recipe has pkgconfig in build requirements when openjpeg option is enabled, and pkgconfig files are generated by PkgConfigDeps, so why would it fail without this patch? I can try to test but AFAIR it was working fine.

SpaceIm replied:
I've tested and previous patch works like a charm, openjpeg is always found & injected

theirix replied:
Yes, my bad. Previously openjpeg was located using pkgconfig. So hardcoded openjpeg's version paths do not affect already found openjpeg.

Why I have moved to cmake-only openjpeg is to fix MSVC build (I don't have a clue why it is failed). 1.83.0 and pre-1.83.0 Leptonica have different resolution order for openjpeg: pkg-config then cmake vs cmake vs pkg-config. I think it is easier for us just to use CMake. How do you think?

theirix commented:
By the way, "ERROR: Package 'zlib/1.2.13' not resolved: Unable to find 'zlib/1.2.13' in remotes" in a Conan v2 automatic comment seems strange.

prince-chrismc commented on recipes/leptonica/all/conanfile.py:
```diff
@@ -139,47 +133,69 @@ def _patch_sources(self):
             replace_in_file(self, cmakelists_src, "if (GIF_LIBRARIES)", "if(0)")
-            replace_in_file(self, cmake_configure, "if (GIF_FOUND)", "if(0)")
+            if Version(self.version) >= "1.83.0":
+                replace_in_file(self, cmake_configure, "if(GIF_FOUND)", "if(0)")
```
I would love to see patches for this project in the future 🙏
